File: web/lib/dedupe.ts
// Query-time near-duplicate detection for the review queue. Surfaces "possible
// duplicate of deal N" as a hint for the reviewer to decide; never auto-merges.
// Computed at read time (not stored) so it stays correct as deals are approved,
// edited, or expired, and so it naturally spans sources (the same promo posted to
// Instagram and a website).

export type DupeRow = {
  id: number;
  chain_id: number | null;
  deal_type: string | null;
  discount_value: number | null;
  valid_from: string | null; // 'YYYY-MM-DD' or null (open-ended)
  valid_to: string | null;
  headline: string | null;
  raw_capture_id: number | null;
  scope?: string | null;
};

export type DuplicateHint = { id: number; headline: string };

// Headline-similarity gates. A shared, equal discount_value is already a strong
// identity signal, so a valued pair needs only weak headline agreement; a
// null-value pair (freebie/bogo/loyalty carry no number) leans entirely on the
// headline, so it must clear a higher bar. This is what separates "first 30"
// from "first 300" when neither has a discount_value.
export const SIM_VALUED = 0.35;
export const SIM_NULL_VALUE = 0.6;

export function normHeadline(s: string | null): string {
  return (s ?? "").toLowerCase().replace(/[^a-z0-9 ]+/g, " ").replace(/\s+/g, " ").trim();
}

function bigrams(s: string): Map<string, number> {
  const m = new Map<string, number>();
  for (let i = 0; i < s.length - 1; i++) {
    const g = s.slice(i, i + 2);
    m.set(g, (m.get(g) ?? 0) + 1);
  }
  return m;
}

// Dice coefficient over character bigrams: 1.0 identical, 0 disjoint.
export function similarity(a: string, b: string): number {
  if (a === b) return a.length ? 1 : 0;
  if (a.length < 2 || b.length < 2) return 0;
  const A = bigrams(a);
  const B = bigrams(b);
  let inter = 0;
  let total = 0;
  for (const [g, n] of A) {
    total += n;
    const bn = B.get(g);
    if (bn) inter += Math.min(n, bn);
  }
  for (const [, n] of B) total += n;
  return (2 * inter) / total;
}

// null chain_id is a wildcard: an unattributed deal can still duplicate a
// chained one. (The 19 legacy Instagram deals have null chain_id, so without
// this they could never match each other.)
function chainCompatible(a: number | null, b: number | null): boolean {
  return a == null || b == null || a === b;
}

// Two deals for different cities are not the same deal, however identical their
// wording. A national chain runs one promo in several cities and posts it once
// per city, so without this the Melbourne and Canberra copies look like perfect
// duplicates - and merging the Melbourne one INTO the Canberra one would hide a
// real, actionable deal behind a row the dashboard refuses to publish.
// 'unknown' stays compatible with everything: it means we do not know, and
// blocking on it would strand exactly the reposts this matcher exists to catch.
function scopeCompatible(a: string | null | undefined, b: string | null | undefined): boolean {
  if (!a || !b) return true;
  if (a === "unknown" || b === "unknown") return true;
  return a === b;
}

function valueCompatible(a: number | null, b: number | null): boolean {
  if (a == null && b == null) return true;
  if (a == null || b == null) return false;
  return a === b;
}

// null valid_from = open start (-infinity), null valid_to = open end (+infinity).
export function windowsOverlap(
  aFrom: string | null,
  aTo: string | null,
  bFrom: string | null,
  bTo: string | null
): boolean {
  const startOk = aFrom == null || bTo == null || aFrom <= bTo;
  const endOk = bFrom == null || aTo == null || bFrom <= aTo;
  return startOk && endOk;
}

export function isPossibleDuplicate(a: DupeRow, b: DupeRow): boolean {
  if (a.id === b.id) return false;
  // Same post yields multiple offers legitimately (a bogo and a step-goal
  // freebie in one caption); those are not duplicates of each other.
  if (a.raw_capture_id != null && a.raw_capture_id === b.raw_capture_id) return false;
  if (!a.deal_type || a.deal_type !== b.deal_type) return false;
  if (!chainCompatible(a.chain_id, b.chain_id)) return false;
  if (!scopeCompatible(a.scope, b.scope)) return false;
  if (!valueCompatible(a.discount_value, b.discount_value)) return false;
  if (!windowsOverlap(a.valid_from, a.valid_to, b.valid_from, b.valid_to)) return false;
  const bothValued = a.discount_value != null && b.discount_value != null;
  const threshold = bothValued ? SIM_VALUED : SIM_NULL_VALUE;
  return similarity(normHeadline(a.headline), normHeadline(b.headline)) >= threshold;
}

export function findDuplicatesFor(target: DupeRow, pool: DupeRow[]): DuplicateHint[] {
  return pool
    .filter((c) => isPossibleDuplicate(target, c))
    .map((c) => ({ id: c.id, headline: c.headline ?? "" }));
}

File: ingest/dedupe.py
"""Duplicate scoring for the nightly merge job.

This is a Python port of the review-queue matcher in web/lib/dedupe.ts. KEEP THE
TWO IN SYNC: the review queue flags possible duplicates with the TS version, and
this one decides which flagged pairs the merge job actually collapses. A drift
between them would merge pairs the UI never surfaced, or vice versa. The
similarity function and thresholds are intentionally identical to the TS file.
"""
import re

# Headline-similarity gates, identical to web/lib/dedupe.ts. Valued pairs (equal
# non-null discount_value) need only weak headline agreement because the value is
# already strong identity; null-value pairs would need the higher bar, but the
# merge job never auto-merges them anyway (see merge_candidate).
SIM_VALUED = 0.35
SIM_NULL_VALUE = 0.6

_NON = re.compile(r"[^a-z0-9 ]+")
_WS = re.compile(r"\s+")


def norm_headline(s):
    s = _NON.sub(" ", (s or "").lower())
    return _WS.sub(" ", s).strip()


def _bigrams(s):
    m = {}
    for i in range(len(s) - 1):
        g = s[i : i + 2]
        m[g] = m.get(g, 0) + 1
    return m


def similarity(a, b):
    """Dice coefficient over character bigrams: 1.0 identical, 0 disjoint."""
    if a == b:
        return 1.0 if a else 0.0
    if len(a) < 2 or len(b) < 2:
        return 0.0
    A, B = _bigrams(a), _bigrams(b)
    inter = total = 0
    for g, n in A.items():
        total += n
        bn = B.get(g)
        if bn:
            inter += min(n, bn)
    for n in B.values():
        total += n
    return (2 * inter) / total


def scope_compatible(a, b):
    """Two deals for different cities are not the same deal, however identical
    their wording. Mirrors scopeCompatible in web/lib/dedupe.ts. 'unknown' stays
    compatible with everything: it means we do not know, and blocking on it would
    strand exactly the reposts this matcher exists to catch."""
    if not a or not b:
        return True
    if a == "unknown" or b == "unknown":
        return True
    return a == b


def windows_overlap(a_from, a_to, b_from, b_to):
    """null valid_from = open start (-inf), null valid_to = open end (+inf)."""
    start_ok = a_from is None or b_to is None or a_from <= b_to
    end_ok = b_from is None or a_to is None or b_from <= a_to
    return start_ok and end_ok


def merge_candidate(a, b):
    """Strict AUTO-MERGE gate (stricter than the review flag). Merge only when all
    hold: same non-null chain_id, same deal_type, equal non-null discount_value,
    overlapping validity window, headline similarity above SIM_VALUED. Two offers
    from the SAME post are never merged. NULL discount_value on either side is
    never auto-merged (the freebie over-flag case is left for the review queue).
    a and b are dicts with those keys. Does not gate on model confidence."""
    if a.get("raw_capture_id") is not None and a.get("raw_capture_id") == b.get("raw_capture_id"):
        return False
    if a.get("chain_id") is None or a.get("chain_id") != b.get("chain_id"):
        return False
    if not scope_compatible(a.get("scope"), b.get("scope")):
        return False
    if not a.get("deal_type") or a.get("deal_type") != b.get("deal_type"):
        return False
    if a.get("discount_value") is None or b.get("discount_value") is None:
        return False
    if float(a["discount_value"]) != float(b["discount_value"]):
        return False
    if not windows_overlap(a.get("valid_from"), a.get("valid_to"), b.get("valid_from"), b.get("valid_to")):
        return False
    return similarity(norm_headline(a.get("headline")), norm_headline(b.get("headline"))) >= SIM_VALUED
